collect_app/src/main/java/org/odk/collect/android/utilities/TimerLogger.java

```diff
@@ -0,0 +1,330 @@
+    private static AsyncTask saveTask = null;
+    private ArrayList<Event> events = null;
+    private String filename = null;
+    private File timerlogFile = null;
+    private long surveyOpenTime = 0;
+    private long surveyOpenElapsedTime = 0;
+    private boolean timerEnabled = false;              // Set true of the timer logger is enabled
+
+
+    public TimerLogger(File instanceFile, SharedPreferences sharedPreferences, FormController formController) {
+
+        /*
+         * The timer logger is enabled if:
+         *  1) The meta section of the form contains a logging entry
+         *      <orx:audit />
+         *  2) And logging has been enabled in the device preferences
+         */
+        boolean loggingEnabledInForm = formController.getSubmissionMetadata().audit;
+        boolean loggingEnabledInPref = sharedPreferences.getBoolean(
+                AdminKeys.KEY_TIMER_LOG_ENABLED, false);
+        timerEnabled = loggingEnabledInForm && loggingEnabledInPref;
+
+        if (timerEnabled) {
+            filename = "audit.csv";
+            if (instanceFile != null) {
+                File instanceFolder = instanceFile.getParentFile();
+                timerlogFile = new File(instanceFolder.getPath() + File.separator + filename);
+            }
+            events = new ArrayList<Event>();
+        }
+    }
+
+
+    public void setPath(String instancePath) {
+        if (timerEnabled) {
+            timerlogFile = new File(instancePath + File.separator + filename);
+        }
+    }
+
+    /*
+     * Log a new event
+     */
+    public void logTimerEvent(EventTypes eventType,
+                              int fecType,
+                              TreeReference ref,
+                              boolean advancingPage,
+                              boolean writeImmediatelyToDisk) {
+
+        if (timerEnabled) {
+
+            // Calculate the time and add the event to the events array
+            long start = getEventTime();
+
+            // Set the node value from the question reference
+            String node = ref == null ? "" : ref.toString();
+            if (node != null && eventType == EventTypes.FEC
+                    && (fecType == FormEntryController.EVENT_QUESTION
+                    || fecType == FormEntryController.EVENT_GROUP)) {
+                int idx = node.lastIndexOf('[');
+                if (idx > 0) {
+                    node = node.substring(0, idx);
+                }
+            }
+
+            Event newEvent = new Event(start, eventType, fecType, node, advancingPage);
+
+            /*
+             * Ignore the event if we are already in an interval view event or have jumped
+             * This can happen if the user is on a question page and the page gets refreshed
+             * The exception is hierarchy events since they interrupt an existing interval event
+             */
+            if (newEvent.isIntervalViewEvent()) {
+                for (int i = 0; i < events.size(); i++) {
```

Comment:
I would prefer to use foeach loop:

> for (Event event : events) {
                    if (event.isIntervalViewEvent() && !event.endTimeSet) {
                        return;
                    }
                }